R/pool.R

```diff
@@ -45,118 +53,178 @@ Pool <- R6Class("Pool",
         }
 
-        freeEnv <- private$freeObjects
         ## see if there's any free objects
+        freeEnv <- private$freeObjects
         if (length(freeEnv) > 0) {
-          ## get first free object we find
-          id <- ls(freeEnv)[[1]]
+          id <- ls(freeEnv)[[1]]  ## get first free object we find
           object <- freeEnv[[id]]
-
-          taskHandle <- attr(object, "reapTaskHandle", exact = TRUE)
-          if (!is.null(taskHandle)) {
-            attr(object, "reapTaskHandle") <- NULL
-            # Cancel the reap task.
-            taskHandle()
-          }
+          private$cancelReapTask(object)  ## cancel reap task if it exists
 
         } else {
           ## if we get here, there are no free objects
           ## and we must create a new one
           object <- private$createObject()
-          id <- attr(object, "id", exact = TRUE)
-        }
-        ## activate object and return it
-        private$changeObjectStatus(id, object, "free", "taken")
-        onActivate(object)
-        if (!onValidate(object)) {
-          stop("Object activation was not successful")
+          id <- attr(object, "..metadata", exact = TRUE)$..id
         }
 
+        ..metadata <- attr(object, "..metadata", exact = TRUE)
+        naiveScheduler$protect({
+          ## stop recurring validation
+          if (!is.null(..metadata$..stopValHandle)) {
+            ..metadata$..stopValHandle()
+          }
+          private$checkValid(object) ## calls onActivate and onValidate
+        })
+
+        private$changeObjectStatus(id, object, "free", "taken")
         return(object)
       })
     },
-    ## passivates the object and returns it back to
-    ## the pool (possibly destroys the object if the
-    ## number of total object exceeds the maximum)
-    release = function(id, object) {                      #### WHAT TO DO IF THIS THROWS AN ERROR???
-      protectDefaultScheduler({
-        freeEnv <- private$freeObjects
-        onPassivate(object)
-        private$changeObjectStatus(id, object, "taken", "free")
 
-        taskHandle <- scheduleTask(self$idleTimeout, function() {
-          if (self$counters$free + self$counters$taken > self$minSize) {
-            private$changeObjectStatus(id, object, "free", NULL)
-            private$destroyObject(object)
-          }
+    ## passivates the object and returns it back to the pool
+    ## (sets up task to destroy the object if the number of
+    ## total objects exceeds the minimum)
+    release = function(id, object) {
+      ..metadata <- attr(object, "..metadata", exact = TRUE)
+      if (..metadata$..state == "free") {
+        stop("This object was already returned to the pool.")
+      }
+      if (is.null(..metadata) || !..metadata$..valid) {
+        stop("Invalid object.")
+      }
+      ## immediately destroy object if pool has already been closed
+      if (!self$valid) {
+        private$changeObjectStatus(id, object, "taken", NULL)
```

Comment:
Never mind... getting rid of the `from` arg from `changeObjectStatus` makes this discussion moot